mono/utils/mono-codeman.c

```diff
@@ -184,6 +190,47 @@ mono_code_manager_install_callbacks (const MonoCodeManagerCallbacks* callbacks)
 	code_manager_callbacks = callbacks;
 }
 
+static
+int
+mono_codeman_get_allocation_type (MonoCodeManager const *cman)
+{
+#ifdef FORCE_MALLOC
+	return CODE_FLAG_MALLOC;
+#else
+	return cman->dynamic ? CODE_FLAG_MALLOC : CODE_FLAG_MMAP;
+#endif
+}
+
+/**
+ * mono_code_manager_new_internal
+ *
+ * Returns: the new code manager
+ */
+static
+MonoCodeManager*
+mono_code_manager_new_internal (gboolean dynamic)
+{
+	MonoCodeManager* cman = g_new0 (MonoCodeManager, 1);
+	if (cman) {
+		cman->dynamic = dynamic;
+#if _WIN32
+		// It would seem the heap should live and die with the codemanager,
+		// but that was failing, so try a global.
+		if (mono_codeman_get_allocation_type (cman) == CODE_FLAG_MALLOC && !mono_code_manager_heap) {
```

Comment:
Arguable. I really don't like explicit global initialization, and this isn't that complicated.
As long as, esp, there is no shutdown.

Think of your favorite libraries -- `fopen`, `open`, `malloc`, `printf` -- and find where you explicitly initialize them -- you don't. They "just work". That is how everything can/should be, including here.

Reply:
In this case we already have the pattern in place, mono_code_manager_init and mono_code_manager_cleanup called during init and shutdown so thought it would make sense to simplify it and do HeapCreate/HeapFree in those methods. I see that some things are initialized as needed, valloc_mutex/valloc_freelists, but that doesn't look thread safe, but works since it called during init, when only one thread is active. This init code will more likely be called when multiple threads are active (potentially calling into this method). I could however agree that lazy initialize of the heap could make sense in this case since we don't use it that frequent, so instead of always paying the price for the heap we will only pay the price when needed. We could however destroy it as part of the mono_code_manager_cleanup. Maybe doing the code a little cleaner and move the heap create/destroy in separate functions?

Reply:
The structure is analogous to the dlmalloc code I think.
Doing the cleanup would leak differently and I fear expose bugs.
Revisit later?
I have at least three approaches to try here that all remove dlmalloc.
This seems the most faithful to the current structure.

Reply:
OK, maybe it will look nicer if we move the heap specific functions that now have different implementations on different platforms into static methods, reducing the number of ifdefs within existing methods.

Reply:
I can try. Making the changes the way I did I contend is an easier to read diff, and easier to see the correctness of long term, or at least, easier to see that the platforms are similar/equivalent.

If I split things up, it might be harder to see they are the same.
But I can try that.

On the other hand, this might be a stepping stone, to replacing codeman on Win32 even, in which case it'll probably be more like you ask.

Reply:
I looked into this today and disagree strongly, mostly. The vast bulk of the code is shared, and only a few lines here and there are ifdef. If you want to factor out those lines into new functions with separate implementations, ok.

Specifically maybe:
`mono_code_manager_new_heap`:
empty for Unix, something for Win32


`mono_code_manager_free`:
`dlfree` for Unix
`HeapFree` for Win32
This occurs twice, so that would help.

`mono_code_manager_malloc`:
`dlmemalign` for Unix
`HeapAlloc` for Win32


Is that you what you meant? Introduce functions at each ifdef point, not ifdef'ing the existing function names? That I can do.

Reply:
Yes, that was my point, instead of having for example:

```
#if _WIN32
			void* const heap = mono_code_manager_heap;
			g_assert (heap);
			HeapFree (heap, 0, dead->data);
#else
			dlfree (dead->data);
#endif
```

having a function:

static
void mono_code_manager_free_internal

and put #ifdef:ed code inside that, or #ifdef around the whole function giving one on windows and another on none windows using dlmalloc etc, moving the specific platform implementation out from the "generic" methods, making them cleaner.

Reply:
Sure, not a big deal, but ok, will do.
Otherwise PR ok?

Reply:
No big deal, but nice. Apart from that PR looks ok.

Reply:
It is nicer, agreed. I only did malloc and free, I left init asis.

It is possible we can prune far more on Win32 here, but this is a good first step, and ok last step imho.